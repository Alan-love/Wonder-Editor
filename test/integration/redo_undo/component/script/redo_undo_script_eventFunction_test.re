open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("redo_undo: script component->event function", () => {
    let sandbox = getSandboxDefaultVal();

    let _createCurrentSceneTreeNodeInspectorSnapshotAndMatch = () => {
      MainEditorSceneTreeTool.Select.selectGameObject(
        ~gameObject=GameObjectTool.unsafeGetCurrentSceneTreeNode(),
        (),
      );
      BuildComponentTool.buildScriptComponent(
        ~script=GameObjectTool.getCurrentSceneTreeNodeScript(),
        (),
      )
      |> ReactTestTool.createSnapshotAndMatch;
    };

    beforeEach(() => {
      sandbox := createSandbox();
      MainEditorSceneTool.initState(~sandbox, ());
      MainEditorSceneTool.createDefaultScene(
        sandbox,
        MainEditorSceneTool.setFirstCubeToBeCurrentSceneTreeNode,
      );

      MainEditorInspectorAddComponentTool.addScriptComponent();
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    describe("test add script event function", () => {
      let _simulate = () => {
        let script = GameObjectTool.getCurrentSceneTreeNodeScript();
        let assetTreeData =
          MainEditorAssetTreeTool.BuildAssetTree.buildEmptyAssetTree();
        let addedNodeId = MainEditorAssetIdTool.getNewAssetId();
        MainEditorAssetHeaderOperateNodeTool.addScriptEventFunction();

        MainEditorScriptEventFunctionTool.addScriptEventFunction(
          ~script,
          ~send=SinonTool.createOneLengthStub(sandbox^),
          (),
        );
      };

      describe("test undo operate", () =>
        describe("test undo one step", () =>
          test("step which from first to zero", () => {
            _simulate();

            RedoUndoTool.undoHistoryState();

            _createCurrentSceneTreeNodeInspectorSnapshotAndMatch();
          })
        )
      );

      describe("test redo operate", () =>
        describe("test redo one step", () =>
          test(
            "undo step which from first to zero, redo step which from zero to first",
            () => {
            _simulate();

            RedoUndoTool.undoHistoryState();
            RedoUndoTool.redoHistoryState();

            _createCurrentSceneTreeNodeInspectorSnapshotAndMatch();
          })
        )
      );
    });

    describe("test change script event function", () => {
      let _simulate = () => {
        let script = GameObjectTool.getCurrentSceneTreeNodeScript();

        MainEditorAssetHeaderOperateNodeTool.addScriptEventFunction();
        MainEditorAssetHeaderOperateNodeTool.addScriptEventFunction();

        MainEditorScriptEventFunctionTool.addScriptEventFunction(
          ~script=GameObjectTool.getCurrentSceneTreeNodeScript(),
          ~send=SinonTool.createOneLengthStub(sandbox^),
          (),
        );

        let targetScriptEventFunctionNodeId =
          MainEditorScriptEventFunctionTool.getUnUsedScriptEventFunctionNodeIds(
            script,
          )
          |> StateLogicService.getStateToGetData
          |> ArrayService.unsafeGetFirst;

        MainEditorScriptEventFunctionTool.handleChangeScriptEventFunction(
          ~script,
          ~currentScriptEventFunctionNodeId=
            MainEditorScriptTool.getScriptAllEventFunctionNodeIds(script)
            |> StateLogicService.getStateToGetData
            |> ArrayService.unsafeGetFirst,
          ~targetScriptEventFunctionNodeId,
          ~send=SinonTool.createOneLengthStub(sandbox^),
          (),
        );
      };

      describe("test undo operate", () =>
        describe("test undo one step", () =>
          test("step which from first to zero", () => {
            _simulate();

            RedoUndoTool.undoHistoryState();

            _createCurrentSceneTreeNodeInspectorSnapshotAndMatch();
          })
        )
      );

      describe("test redo operate", () =>
        describe("test redo one step", () =>
          test(
            "undo step which from first to zero, redo step which from zero to first",
            () => {
            _simulate();

            RedoUndoTool.undoHistoryState();
            RedoUndoTool.redoHistoryState();

            _createCurrentSceneTreeNodeInspectorSnapshotAndMatch();
          })
        )
      );
    });

    describe("test remove script event function", () => {
      let _simulate = () => {
        let script = GameObjectTool.getCurrentSceneTreeNodeScript();
        let assetTreeData =
          MainEditorAssetTreeTool.BuildAssetTree.buildEmptyAssetTree();
        let addedNodeId = MainEditorAssetIdTool.getNewAssetId();
        MainEditorAssetHeaderOperateNodeTool.addScriptEventFunction();

        MainEditorScriptEventFunctionTool.addScriptEventFunction(
          ~script,
          ~send=SinonTool.createOneLengthStub(sandbox^),
          (),
        );

        let eventFunctionName =
          ScriptEventFunctionInspectorTool.getEventFunctionName(addedNodeId)
          |> StateLogicService.getEditorState;

        MainEditorScriptEventFunctionTool.removeScriptEventFunction(
          ~script,
          ~eventFunctionName,
          ~dispatchFunc=SinonTool.createOneLengthStub(sandbox^),
          (),
        );
      };

      describe("test undo operate", () =>
        describe("test undo one step", () =>
          test("step which from first to zero", () => {
            _simulate();

            RedoUndoTool.undoHistoryState();

            _createCurrentSceneTreeNodeInspectorSnapshotAndMatch();
          })
        )
      );

      describe("test redo operate", () =>
        describe("test redo one step", () =>
          test(
            "undo step which from first to zero, redo step which from zero to first",
            () => {
            _simulate();

            RedoUndoTool.undoHistoryState();
            RedoUndoTool.redoHistoryState();

            _createCurrentSceneTreeNodeInspectorSnapshotAndMatch();
          })
        )
      );
    });
  });