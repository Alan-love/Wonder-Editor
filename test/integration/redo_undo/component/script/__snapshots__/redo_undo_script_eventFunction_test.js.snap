// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redo_undo: script component->event function test add script event function test redo operate test redo one step undo step which from first to zero, redo step which from zero to first 1`] = `
<article
  className="wonder-inspector-script"
>
  <article
    className="wonder-inspector-script"
  >
    <div>
      <div
        className="inspector-item"
      >
        <div
          className="item-header"
          title="script event function asset in use"
        >
          Script Event Function
        </div>
        <div
          className="item-content"
        >
          <div
            className="inspector-select"
          >
            <div
              className="select-name"
              onClick={[Function]}
            >
              New Script EventFunction
            </div>
            <div
              className="select-img"
              onClick={[Function]}
            >
              <img
                src="./public/img/select.png"
              />
            </div>
          </div>
        </div>
      </div>
      <button
        className="scriptEventFunction-remove"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
    <button
      className="addable-btn"
      onClick={[Function]}
    >
      Add script event function
    </button>
  </article>
  <article
    className="wonder-inspector-script"
  >
    <button
      className="addable-btn"
      onClick={[Function]}
    >
      Add script attribute
    </button>
  </article>
</article>
`;

exports[`redo_undo: script component->event function test add script event function test undo operate test undo one step step which from first to zero 1`] = `
<article
  className="wonder-inspector-script"
>
  <article
    className="wonder-inspector-script"
  >
    <button
      className="addable-btn"
      onClick={[Function]}
    >
      Add script event function
    </button>
  </article>
  <article
    className="wonder-inspector-script"
  >
    <button
      className="addable-btn"
      onClick={[Function]}
    >
      Add script attribute
    </button>
  </article>
</article>
`;

exports[`redo_undo: script component->event function test change script event function test redo operate test redo one step undo step which from first to zero, redo step which from zero to first 1`] = `
<article
  className="wonder-inspector-script"
>
  <article
    className="wonder-inspector-script"
  >
    <div>
      <div
        className="inspector-item"
      >
        <div
          className="item-header"
          title="script event function asset in use"
        >
          Script Event Function
        </div>
        <div
          className="item-content"
        >
          <div
            className="inspector-select"
          >
            <div
              className="select-name"
              onClick={[Function]}
            >
              New Script EventFunction 1
            </div>
            <div
              className="select-img"
              onClick={[Function]}
            >
              <img
                src="./public/img/select.png"
              />
            </div>
          </div>
        </div>
      </div>
      <button
        className="scriptEventFunction-remove"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
    <button
      className="addable-btn"
      onClick={[Function]}
    >
      Add script event function
    </button>
  </article>
  <article
    className="wonder-inspector-script"
  >
    <button
      className="addable-btn"
      onClick={[Function]}
    >
      Add script attribute
    </button>
  </article>
</article>
`;

exports[`redo_undo: script component->event function test change script event function test undo operate test undo one step step which from first to zero 1`] = `
<article
  className="wonder-inspector-script"
>
  <article
    className="wonder-inspector-script"
  >
    <div>
      <div
        className="inspector-item"
      >
        <div
          className="item-header"
          title="script event function asset in use"
        >
          Script Event Function
        </div>
        <div
          className="item-content"
        >
          <div
            className="inspector-select"
          >
            <div
              className="select-name"
              onClick={[Function]}
            >
              New Script EventFunction
            </div>
            <div
              className="select-img"
              onClick={[Function]}
            >
              <img
                src="./public/img/select.png"
              />
            </div>
          </div>
        </div>
      </div>
      <button
        className="scriptEventFunction-remove"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
    <button
      className="addable-btn"
      onClick={[Function]}
    >
      Add script event function
    </button>
  </article>
  <article
    className="wonder-inspector-script"
  >
    <button
      className="addable-btn"
      onClick={[Function]}
    >
      Add script attribute
    </button>
  </article>
</article>
`;

exports[`redo_undo: script component->event function test remove script event function test redo operate test redo one step undo step which from first to zero, redo step which from zero to first 1`] = `
<article
  className="wonder-inspector-script"
>
  <article
    className="wonder-inspector-script"
  >
    <button
      className="addable-btn"
      onClick={[Function]}
    >
      Add script event function
    </button>
  </article>
  <article
    className="wonder-inspector-script"
  >
    <button
      className="addable-btn"
      onClick={[Function]}
    >
      Add script attribute
    </button>
  </article>
</article>
`;

exports[`redo_undo: script component->event function test remove script event function test undo operate test undo one step step which from first to zero 1`] = `
<article
  className="wonder-inspector-script"
>
  <article
    className="wonder-inspector-script"
  >
    <div>
      <div
        className="inspector-item"
      >
        <div
          className="item-header"
          title="script event function asset in use"
        >
          Script Event Function
        </div>
        <div
          className="item-content"
        >
          <div
            className="inspector-select"
          >
            <div
              className="select-name"
              onClick={[Function]}
            >
              New Script EventFunction
            </div>
            <div
              className="select-img"
              onClick={[Function]}
            >
              <img
                src="./public/img/select.png"
              />
            </div>
          </div>
        </div>
      </div>
      <button
        className="scriptEventFunction-remove"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
    <button
      className="addable-btn"
      onClick={[Function]}
    >
      Add script event function
    </button>
  </article>
  <article
    className="wonder-inspector-script"
  >
    <button
      className="addable-btn"
      onClick={[Function]}
    >
      Add script attribute
    </button>
  </article>
</article>
`;
